omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+#include "drvOmroneip.h"
+#include <sstream>
+
+// This stores information about each communication tag to the PLC.
+// A new instance will be made for each record which requsts to uniquely read/write to the PLC
+struct omronDrvUser_t
+{
+  size_t startIndex;
+  size_t sliceSize;
+  std::string tag;
+  std::string dataType;
+  int32_t tagIndex;
+  size_t dataCounter;
+  std::string pollerName;
+  size_t tagOffset;
+  double timeout;
+  std::string optimisationFlag;
+  size_t strCapacity;
+  bool readFlag;
+};
+
+typedef struct
+{
+  omronDataType_t dataType;
+  const char *dataTypeString;
+} omronDataTypeStruct;
+
+// Supported PLC datatypes
+static omronDataTypeStruct omronDataTypes[MAX_OMRON_DATA_TYPES] = {
+    {dataTypeBool, "BOOL"},
+    {dataTypeInt, "INT"},
+    {dataTypeDInt, "DINT"},
+    {dataTypeLInt, "LINT"},
+    {dataTypeUInt, "UINT"},
+    {dataTypeUDInt, "UDINT"},
+    {dataTypeULInt, "ULINT"},
+    {dataTypeUInt_BCD, "UINT_BCD"},
+    {dataTypeUDInt_BCD, "UDINT_BCD"},
+    {dataTypeULInt_BCD, "ULINT_BCD"},
+    {dataTypeReal, "REAL"},
+    {dataTypeLReal, "LREAL"},
+    {dataTypeString, "STRING"},
+    {dataTypeWord, "WORD"},
+    {dataTypeDWord, "DWORD"},
+    {dataTypeLWord, "LWORD"},
+    {dataTypeUDT, "UDT"},
+    {dataTypeUDT, "TIME"}};
+
+bool iocStarted = false; // Set to 1 after IocInit() which starts pollers
+
+static const char *driverName = "drvOmronEIP"; /* String for asynPrint */
+
+static void readPollerC(void *drvPvt)
+{
+  drvOmronEIP *pPvt = (drvOmronEIP *)drvPvt;
+  pPvt->readPoller();
+}
+
+// this thread runs once after iocInit to optimise the tag map before the thread poller begins
+static int optimiseTagsC(void *drvPvt)
+{
+  drvOmronEIP *pPvt = (drvOmronEIP *)drvPvt;
+  while (!iocStarted)
```

Comment:
At a glance and judging by variable name `iocStarted` I suspect you're waiting for `iocInit()` to complete. However, there are hooks/events for the various stages of IOC initialisation which I would recommend instead of a polling loop. See https://docs.epics-controls.org/en/latest/appdevguide/IOCInit.html#inithooks The hook you're after is probably `initHookAfterIocRunning`

Reply:
Ah, I see now that you *are* using the initHook :+1:

Then what I would suggest is to use an epicsEvent to signal that you're started/initialized instead of going around a while loop. The epicsEvent.h header has this API and the old Application Developers Guile describe it on p. 289 (I couldn't easily find it in the modern EPICS docs) 

https://epics.anl.gov/base/R3-14/12-docs/AppDevGuide.pdf

Reply:
The reason iocStarted was not an epicsEvent, is because either the iocStarted flag or the iocStartedEvent would have to be global due to the required format of the hook function. (static void myInitHookFunction(initHookState state))

Looking at it now however, I dont think there is really any difference between having a global bool and a global epicsEvent, so I will change this.

Reply:
Is there any way of finding out if an event has fired in the past? Elsewhere in the code I need to check if the event has fired in the past

Reply:
After I have triggered and then waited for the signal elsewhere, I could potentially rearm the event and then check that the event status is full? This would mean that the signal had either been fired and received once or just fired. It doesnt feel great though.